common/src/main/java/org/conscrypt/SSLUtils.java

```diff
@@ -577,5 +576,21 @@ private static int unsignedShort(short s) {
         return result;
     }
 
+    // Flag file created by CtsConscryptFdSocketTestCases.  This is a temporary measure
+    // so CTS can test both implementations easily until FdSocket is retired.
+    // Name chosen so that it is unlikely to exist by accident on other platforms.
+    // If this file exists, always use fd-based socket.  Otherwise default to true but
+    // allow overriding by system property.
+    // TODO(prb): Consider gating on target API level on Android.
+    private static final String CTS_FDSOCKET_FILE
+        = "/data/local/tmp/ctslibcore/org.conscrypt.use_fd_socket";
+    private static boolean useEngineSocketByDefault() {
+        if (new File(CTS_FDSOCKET_FILE).exists()) {
+            return false;
```

Comment:
Not sure I'd want to deploy this in production, what happens if this file is created on a non-android device?

Could we produce loud logging (WARN or above) if the test override is enabled.
I'm curious why jvm sysprop is insufficient, would an envar be more feasible?

Can we avoid this check in situations where we're confident tests aren't running?
For example, if asserts aren't enabled this can be ignored:
```java
@SuppressWarnings("AssertWithSideEffects")
private static boolean assertsEnabled() {
    boolean enabled = false;
    assert enabled = true;
    return enabled;
}
```
or alternatively we can attempt to load a sentinel class that's only expected to be available in testing (CtsConscryptFdSocketTestCases?).

Reply:
Thanks for the review and the pertinent questions and suggestions!

Some background:-
Currently Conscrypt contains two implementations of SSLSocket. One ([ConscryptFileDescriptorSocket](https://github.com/google/conscrypt/blob/master/common/src/main/java/org/conscrypt/ConscryptFileDescriptorSocket.java)) operates directly on the file descriptor of the underlying socket and contains a lot of switches into native code and duplicate logic from SSLEngine.  The other ([ConscryptEngineSocket](https://github.com/google/conscrypt/blob/master/common/src/main/java/org/conscrypt/ConscryptEngineSocket.java)) is implemented using an SSLEngine with added logic to move data to and from the underlying socket.  Long term we want to shift all platforms to the engine-based socket and retire the fd-based one to remove complexity and duplication.

In the shorter term, the default outside of the Android platform remains the fd-based socket as we don't have full feature parity yet (e.g. around unblocking threads waiting in system calls when another thread closes the SSLSocket). On [Android 11 developer previews](https://developer.android.com/preview), the Android platform default is the engine-based socket and we're reasonably confident that there are no show-stopping app-compat issues arising from the switch.

Slight gotcha:  my understanding is that on openjdk13, the fd-based implementation doesn't work due to changes in the Socket implementation so apps targeting openjdk13 have to use the engine-based implementation.  Apps can switch the global default either by setting a jvm property or by calling [Conscrypt.setUseEngineSocketByDefault()](https://github.com/google/conscrypt/blob/master/common/src/main/java/org/conscrypt/Conscrypt.java#L271) or they can set it for a individual socket factories via Conscrypt.setUseEngineSocket().

Obviously whilst the switchover is in progress, we want to ensure that both implementations are fully unit tested.  During non-Android-platform development that's fairly straightforward as we have Gradle rules that run all the tests for each implementation.  However Android platform continuous testing uses [CTS](https://source.android.com/compatibility/cts) and its own builds rules, and contains tests that live outside of Conscrypt and exercise the public APIs such as javax.net.ssl.* so the purpose of this change is to enable  continuous testing of both implementations in that environment.  

Perhaps surprisingly, there is currently no way in CTS to set arbitrary JVM properties or environment variables for a test.  And because Conscrypt is part of the core platform libraries on Android it can't use other mechanisms such as Android system properties as those are only available to framework-level code.  So the options came down to: implement a custom test runner for this purpose (which is maybe more engineering work than a temporary solution merits) or crowbar an [existing mechanism for running shell commands](https://cs.android.com/android/platform/superproject/+/master:cts/tests/libcore/luni/AndroidTest.xml;l=25) to create flag files before running a test. And I'd rather take the simpler path.  :) 

So... The impact of this file accidentally existing by accident:-

- On any platform using the Conscrypt.set* APIs I mentioned above, no impact
- For apps using the Android platform APIs, it will flip the default from engine-based socket to fd-based, which is suboptimal but shouldn't cause any issues.
- For non-Android-platform apps which don't try and change the default implementation, no impact
- For non-Android-platform apps which set the default to engine-based socket using JVM properties, potential breakage (e.g. if running on openjdk 13)

As you suggest, the last one is not an acceptable risk and so I'll update the change to only look for that flag file if running under CTS.  It's also possible that when I try and make the CTS changes to enable the extra tests that I'll get pushback forcing me to go the custom test runner route anyway.

Reply:
Thank you for such a thorough response, and for looking into other options. I really appreciate both!

Reply:
FWIW this thread also gave me the nudge I needed to realise I can probably do this without touching Conscrypt at all and without the effort of getting a new test runner approved, so thanks again :)